[1단계 - 출석 구현] 노랑(노주희) 미션 제출합니다.

yesjuhee opened the pull request:
안녕하세요! 리뷰 잘 부탁드립니다 😄  

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요? => `"2024-12-13"과 같이 main 실행시 arguments를 넣어줘야 합니다`
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3900


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

객체지향 생활체조 요구사항보다는 미션에 작성된 요구사항에 집중해서 구현했기 때문에 신경쓰지 못한 부분이 많습니다.
indent는 2를 최대로 사용하였고 `AttendanceRecord` 클래스에서 3개의 인스턴스 변수를 사용했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 1. 모든 원시값을 포장하기
모든 원시값을 포장하라는 요구사항에 따라서 `Crew` 객체를 생성했습니다.
`Crew` 객체는 `CrewAttendanceRecords`에서 출석 기록의 Key로서만 이용되기 때문에 별도의 메서드가 위치하지 않고 생성자와 게터만 존재하기 때문에 Record로 변경하여 작성했습니다.
이러한 상황을 보면서 "굳이 이 원시값 포장이 필요했을까?"라는 생각이 들었습니다.
이에 대한 엘리의 생각은 어떠신지 궁금합니다!

### 2. 시작 날짜 입력 (`DateGenerator`)
미션 요구사항은 12월 한 달 동안 사용할 프로그램을 구현할 것을 이야기하고 있고, 미션을 구현하고 있는 지금은 2월이기 때문에
main 메서드를 실행할 때 `args`를 통해서 기준이 될 현재 날짜를 입력 받도록 구현하고 있습니다.
이를 위해 `DateGenerator`라는 인터페이스를 정의했고, `AttendanceController`에서 람다식을 이용해 인터페이스를 구현해서 입력된 인자 값에 맞는 `LocalDate`를 반환하도록 작성했습니다.
인터페이스가 적절한 방식으로 작성됐는지 피드백을 받고 싶습니다!

### 3. 팩터리 메서드 패턴 (`CsvParsingGenerator`)
모든 크루의 출석 기록을 저장하고 있는 `CrewAttendanceRecords`를 위해서 csv 파일을 읽어온 후 적절한 객체를 생성해서 Map을 만들어주는 로직이 필요합니다. 
`CrewAttendanceRecords`가 생성 로직을 모두 가지고 있을 경우 책임이 과도하게 부여되고 테스트가 어려워지는 점을 고려해 생성 로직을 분리한 `CrewAttendanceRecordsGenerator` 인터페이스와 구현체인 `CsvParsingGenerator`를 만들었습니다.
크루들과 로또 미션 피드백을 하면서 "팩터리 메서드 패턴"을 알게 되어 해당 패턴을 적용해 본 결과인데요.
이 패턴을 사용해 본 것이 처음이라 잘 작성이 된건지, 개선사항은 없는지 궁금합니다!

### 4. 정적 팩터리 메서드
`AttendanceRecord`가 상황에 따라 다양한 인자를 받아서 생성이 되기 때문에 처음에 여러개의 생성자를 작성했습니다.
이 부분을 리팩터링하면서 정적 팩터리 메서드를 사용하도록 모두 변경했습니다.
이 부분도 잘 작성이 된건지, 개선 사항은 없는지 여쭤보고 싶습니다.

jbilee committed:
docs(readme): 기능 요구 명세 작성

- 애플리케이션 구현에 필요한 기능 목록 작성

jbilee committed:
chore: 출석 기록 .csv 파일 생성

jbilee committed:
feat(day, attendance): 출석 관리 기능 구현

- 입력한 출석 시간에 따라 출석, 지각 또는 결석으로 처리한다
- 등교 시간이 캠퍼스 운영 시간이 아닌 경우 예외를 발생시킨다

jbilee committed:
feat(disciplinary_status): 제적 위험자 관리 기능 구현

- 누적 지각 및 결석 횟수에 따라 제적 위험자로 관리한다

jbilee committed:
feat(attendance): 출석 확인 관련 기능 구현

- 등교일이 아닐 경우(주말, 공휴일) 예외를 발생시킨다

jbilee committed:
feat(attendance_record): 출석 관리 기능

- 프로그램 시작 시 출석 기록 파일에서 출석 기록을 불러온다.

jbilee committed:
feat(attendance_record): 출석 수정 기능

- 출석 기록이 없는 날짜를 입력한 경우 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 수정 기능

- 출석 확인을 수정하려면 닉네임, 수정하려는 날짜, 등교 시간을 입력하여 기록을 수정할 수 있다.

jbilee committed:
feat(attendance_record): 출석 수정 기능

- 기록이 없는 닉네임을 입력하면 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 기록 확인 기능

- 기록이 없는 닉네임을 입력하면 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 확인 기능

- 기록이 없는 닉네임을 입력하면 예외를 발생시킨다.

jbilee committed:
feat(attendance_record): 출석 확인 기능

- 이미 출석을 하였는데 다시 출석 확인을 하는 경우 예외를 발생시킨다.

jbilee committed:
chore(global): 패키지 생성 및 파일 이동

- `domain` 패키지 생성
- 도메인에 해당하는 클래스를 `domain` 패키지로 이동
- `Application` 생성

jbilee committed:
feat(menu): 메뉴 기능

- 날짜와 메뉴 리스트를 안내한다.

jbilee committed:
feat(menu): 메뉴 기능

- 종료를 선택하면 프로그램을 종료한다.

jbilee committed:
feat(menu): 메뉴 기능

- 사용자가 입력한 값에 따라 해당 기능 실행한다.

jbilee committed:
feat(check_in): 출석 확인 기능

- 시간 입력 형식이 잘못된 경우 예외를 발생시킨다.

jbilee committed:
feat(check_in): 출석 확인 기능

- 닉네임과 등교 시간을 입력하면 출석할 수 있다.
- 출석 후 출석 기록을 확인할 수 있다.

jbilee committed:
feat(update_attendance): 출석 수정 기능

- 수정 후에는 변경 전과 변경 후의 출석 기록을 확인할 수 있다.
- 시간 입력 형식이 잘못된 경우 예외를 발생시킨다.

jbilee committed:
feat(check_attendance): 출석 기록 확인 기능

- 출석 기록 목록을 출력한다.

jbilee committed:
feat(check_attendance): 출석 기록 확인 기능

- 출결 현황을 출력한다. (출석, 지각, 결석)

jbilee committed:
feat(check_attendance): 출석 기록 확인 기능

- 닉네임을 입력하면 전날까지의 크루 출석 기록을 확인할 수 있다.
- 면담 대상자 여부를 출력한다.

jbilee committed:
feat(check_disciplinary): 제적 위험자 확인

- 전날까지의 크루 출석 기록을 바탕으로 제적 위험자를 파악한다.

jbilee committed:
feat(menu): 메뉴 기능

- 예외가 발생하면 초기 화면으로 돌아간다.

jbilee committed:
refactor(controller) : AttendanceController 리팩터링

- 메뉴 선택 로직 리팩터링
- 상수 추출
- 메서드 순서 변경

jbilee committed:
feat(update_attendance) : 출석 수정 기능

- 날짜 입력 형식이 잘못된 경우 예외를 발생시킨다.

jbilee committed:
refactor(view) : OutputView 리팩터링

- 메소드 순서 변경

jbilee committed:
refactor(domain) : Crew 리팩터링

- 레코드로 변경

jbilee committed:
refactor(domain) : Attendance 리팩터링

- 메소드 순서 변경
- enum 표기 방식 수정

jbilee committed:
refactor(day) : Day 리팩터링

- 사용하지 않는 isHoliday 멤버 변수 삭제
- getDay 메서드 실패시 throw 하도록 수정
- checkHoliday 로직 수정

jbilee committed:
refactor(domain) : DisciplinaryStatus 리팩터링

- 메소드 순서 변경
- 메소드 이름 변경

jbilee committed:
refactor(domain) : AttendanceRecord 리팩터링

- 정적 팩토리 메서드 사용
- 메서드 순서 변경

jbilee committed:
fix(day) : Day 버그 수정

- checkHoliday 메서드 오류 수정

jbilee committed:
refactor(domain) : 중복된 메서드 통합

- getAttendanceCount 메서드 생성

jbilee committed:
refactor(domain) : 메서드 분리

- getDisciplinaryStatus 메서드 생성

jbilee committed:
style(attendance_records) : 스타일 통일

- 줄바꿈 추가

jbilee committed:
refactor(domain) : CrewAttendanceRecords 생성 방식 리팩터링

- 팩토리 메서드 패턴 사용
- CsvParsingGenerator 추가

jbilee committed:
refactor(domain) : CrewAttendanceRecords 리팩터링

- 메서드 분리
- 메서드 순서 변경

jbilee committed:
refactor(global) : 전체적인 코드 리팩터링

- 미사용 메서드 삭제
- 생략된 final 키워드 추가
- 가독성을 위한 라인 분리

jbilee committed:
refactor(domain) : CsvParsingGenerator 리팩터링

- 불필요한 매개변수 삭제

yebink commented on src/main/java/domain/Day.java:
```diff
@@ -0,0 +1,53 @@
+        this.start = start;
+    }
+
+    public static boolean checkHoliday(LocalDate date) {
```
네이밍만 보면, 공휴일인지 여부를 체크할 것 같은데, 공휴일 + 주말 여부를 함께 체크하는군요.

보다 적절한 네이밍으로 바꿔볼까요~?

yebink replied:
boolean 을 리턴할 때에는 `check~` 보다는 `is~` 를 사용하는 것이 일반적입니다 🙂

yesjuhee replied:
오호 감사합니다! 좀 더 적절한 네이밍을 고려해보겠습니다!

yesjuhee replied:
4a79a58

yebink commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public boolean isTardy() {
+        return attendance.equals(Attendance.TARDY);
```
enum 은 보통 `equals` 보다는 `==` 비교를 사용합니다.

```suggestion
        return attendance == Attendance.TARDY;
```

yesjuhee replied:
앗 이 부분을 놓쳤네요 감사합니다!

yesjuhee replied:
7ee6ac1

yebink commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,44 @@
+        return name;
+    }
+
+    private static void validateOpenTime(LocalTime time) {
+        if (time.isBefore(OPEN_HOUR) || time.isAfter(CLOSE_HOUR)) {
+            throw new IllegalArgumentException("[ERROR] 캠퍼스 운영 시간에만 출석이 가능합니다.\n");
+        }
+    }
```
요구사항에 없는 부분까지 구현하셨네요 👍

yebink commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,80 @@
+        this.attendance = Attendance.getAttendanceStatus(Day.getDay(date), time);
+    }
+
+    public static AttendanceRecord parse(String dateTime) {
+        return new AttendanceRecord(LocalDate.parse(dateTime.split(" ")[DATE_INDEX]),
+                LocalTime.parse(dateTime.split(" ")[TIME_INDEX]));
+    }
```
parse 가 AttendanceRecord 의 역할이 맞을까요~?
csv 파일 형식이 `쿠키,2024-12-13 10:08` 에서 `쿠키,2024년 12월 13일 10시08분` 으로 변경되면 domain 까지 수정이 필요한데요!

더 적절한 위치를 고민해보아요 😄

yebink replied:
`dateTime.split(" ")` 은 한 번만 해도 될 것 같네요!

yesjuhee replied:
> parse 가 AttendanceRecord 의 역할이 맞을까요~?
csv 파일 형식이 쿠키,2024-12-13 10:08 에서 쿠키,2024년 12월 13일 10시08분 으로 변경되면 domain 까지 수정이 필요한데요!
> 더 적절한 위치를 고민해보아요 😄

오호 `LocalDate.parse()` 와 같은 역할을 생각했는데 view에 좀 더 가깝다고 볼 수 있겠네요
좋은 리뷰 감사합니다 👍

yesjuhee replied:
csv 파일을 뷰를 통해 읽어오는게 아니어서 우선 문자열 파싱하는 부분을 `CsvParsingGenerator`에 넣었습니다.
`CsvParsingGenerator`가 도메인에 해당하긴 하지만 특정 csv 파일을 읽어오는 역할을 담당하기 때문에 `AttendanceRecord`보다는 `CsvParsingGenerator`가 해당 책임을 가지는 것이 좀 더 적절하다고 판단했습니다!
이 위치가 적절할까요?

d2ef0b5
1ea3cff

yebink commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,72 @@
+
+    public boolean hasRecordOfDate(LocalDate date) {
+        return attendanceRecords.stream()
+                .anyMatch((record) -> record.getDate().equals(date));
```
불필요한 괄호는 제거해보아요~

```suggestion
                .anyMatch(record -> record.getDate().equals(date));
```

yebink replied:
attendanceRecords 가 비어있다면 해당 메서드를 수행할 때 NPE 가 발생할텐데요!
방어 로직을 작성해봅시다~

yesjuhee replied:
> attendanceRecords 가 비어있다면 해당 메서드를 수행할 때 NPE 가 발생할텐데요!

오 중요한 부분인데 놓쳤네요 감사합니다!

yesjuhee replied:
```java
  @Test
  void test() {
      AttendanceRecords attendanceRecords = new AttendanceRecords();
      assertThat(attendanceRecords.hasRecordOfDate(LocalDate.now())).isFalse();
  }
```

위와같이 테스트코드를 작성했을 때 테스트가 통과해서 추가 학습을 해봤습니다.
빈 Stream일 경우 `anyMatch()`는 `false`를 반환하고 `allMatch()`와 `noneMatch()`는 `true`를 반환한다고 합니다!
다행히 지금 코드는 의도대로 동작하네욥
하나 더 배워갑니다 😄

yesjuhee replied:
7ee6ac1

yesjuhee replied:
불필요한 괄호 제거 - e28e945

yebink commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.function.Supplier;
+
+public class AttendanceController {
+    private final static int START_DATE_INDEX = 0;
```
modifier 순서를 지켜주세요 🙂 

```suggestion
    private static final int START_DATE_INDEX = 0;
```

참고 : https://google.github.io/styleguide/javaguide.html#s4.8.7-modifiers

yesjuhee replied:
e28e945

yebink commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,83 @@
+            try {
+                return LocalDate.parse(args[START_DATE_INDEX]);
+            } catch (DateTimeParseException e) {
+                throw new IllegalArgumentException("[ERROR] 프로그램 인수를 YYYY-MM-DD 형식으로 입력해 주세요.");
```
`YYYY-MM-DD` 와 `yyyy-MM-dd` 의 차이에 대해 알아볼까요?

yesjuhee replied:
오오 Date 포멧팅에 대해서 한 번 공부해보겠습니다 🫡

yesjuhee replied:
학습해보니까 아래처럼 정의되어 있네요
- `y` : 년도
- `M` : 월
- `m` : 분
- `D` : 년의 몇 번째 일(1~366)
- `d` : 월의 몇 번째 일(1~31)

새로 배워갑니다!

yesjuhee replied:
80d03a0

이 부분은 리팩터링하면서 2024년 12월의 날짜만 받도록 검증 로직을 변경했습니다!

yebink commented on src/main/java/domain/AttendanceRecords.java:
```diff
@@ -0,0 +1,72 @@
+                .anyMatch((record) -> record.getDate().equals(date));
+    }
+
+    public void fillAbsences(DateGenerator dateGenerator) {
```
크루마다 출석이 없는 날짜에 대해 `결석` 을 기록하신 이유가 무엇인가요~?

크루가 1,000명, 10,000명으로 늘어도 괜찮을까요? 🤔

yesjuhee replied:
요구사항 중 하나라고 판단해서 출석이 없는 날짜에 대해 `결석`으로 기록했습니다.
이 문제가 7기 최종 코테 문제였는데 그때 주어진 `attendances.csv` 기준으로 기록이 없는 크루의 경우에 `3. 크루별 출석 기록 확인
` 기능을 사용했을 때 `결석`으로 표기했기 때문입니다.
또한 `4. 제적 위험자 확인` 기능에서도 기록이 없는 날짜는 결석으로 카운트가 되야하기 때문에 결석으로 채워두는 것이 맞다고 판단했습니다!

yebink replied:
그러한 이유였군요! 😄
null 을 지양하고자 하는 의견에 동의하지만,
임의의 시간을 지정할 경우, 누군가 오해하는 일이 발생할 수도 있지 않을까요?

필요하면 사용하되, NPE가 발생하지 않도록 잘 방어하는 연습을 해보는 것도 중요합니다 👍

yebink replied:
지금과 같은 요구사항에서는 무리 없이 동작하지만, 크루 수가 많아지거나 저장해야 할 날짜가 몇 년치가 되는 경우 메모리에 상당히 많은 결석 데이터가 적재되어야 합니다.
조금 더 확장성을 고려해서, `결석` 에 대한 데이터를 저장하지 않고 출력 시에 계산한다면 좀 더 가벼운 프로그램이 될 수 있을 것 같습니다 🙂

yesjuhee replied:
좋은 접근 방식 제시 감사합니다 👍 
step2에 고려해보도록 하겠습니다!

yebink commented on src/test/java/domain/AttendanceRecordsTest.java:
```diff
@@ -0,0 +1,158 @@
+class AttendanceRecordsTest {
+    @Test
+    @DisplayName("출석 기록이 없는 날짜가 결석으로 기록되었는지 확인한다.")
+    void fillAbsencesTest() {
+        CrewAttendanceRecords crewAttendanceRecords = new CrewAttendanceRecords(new CsvParsingGenerator(), () -> LocalDate.of(2024, 12, 13));
+        Crew crew = new Crew("쿠키");
+        boolean hasRecord = crewAttendanceRecords.hasRecord(crew, LocalDate.of(2024, 12, 12));
+
+        assertThat(hasRecord).isTrue();
+    }
```
테스트에 `// given`, `// when`, `// then` 주석을 달면 가독성을 높일 수 있어요 🙂

yesjuhee replied:
넵 반영해보도록 하겠습니다!

yesjuhee replied:
1ea3cff

yebink commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,78 @@
+        String input = scanner.nextLine();
+        try {
+            int dateNumber = Integer.parseInt(input);
+            return LocalDate.of(2024, 12, dateNumber);
```
노랑이 구현한 프로그램은 arguments 로 날짜를 입력할 수 있게 되어 있는데, 수정은 2024년 12월만 가능하군요!

요구사항에서 좀 더 나아가 수정도 유연한 구조로 변경해볼까요~?

yesjuhee replied:
이 부분은 요구사항에서 <**2024년 12월 한 달 동안** 시범적으로 최소한의 기능을 갖춘 출석 시스템을 개발하여 출석을 체계적으로 관리하기로 하였습니다.> 이라고 명시가 되어있었기 때문에 12월만 작동하는 것을 가정하여 구현했습니다!

`arguments` 로 받는 날짜를 12월만 제한하도록 수정해보겠습니다ㅎㅎ

yesjuhee replied:
기준이 되는 날짜를 2024년 12월의 날짜로만 입력받을 수 있도록 수정했습니다.
이 과정에서 입력받은 인자를 검증하는 책임을 컨트롤러에게 주었습니다.
검증 책임을 컨트롤러가 가지도록 설계를 해본적이 별로 없어서 낯설계 느껴지는데 엘리가 보기에는 어떤지 궁금합니다!

80d03a0

yebink replied:
네 지금은 해당 로직을 가질만한 부분이 controller 뿐이라 크게 이상하진 않습니다 ㅎㅎ 😄

yebink commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Scanner;
+
+public class InputView {
+    public static final String MENU_REGEX = "[1234Qq]";
```
소문자 디테일 👍

yebink commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.Objects;
+
+public record Crew(String name) {
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) return false;
+        Crew crew = (Crew) o;
+        return Objects.equals(name, crew.name);
+    }
```
재정의가 꼭 필요할까요~?

yesjuhee replied:
원시값을 포장할 경우 기본적으로 `hashCode`, `equals`의 재정의가 되어있어야 한다고 생각해서 재정의했습니다.
닉네임이 같은 크루의 경우 같은 객체로 판단해야 한다고 생각했습니다!

yebink replied:
record 는 기본으로 equals() 와 hashCode() 메서드가 자동 생성됩니다.
따라서 재정의는 불필요합니다.
record의 개념에 대해 더 학습해보아요😄

yesjuhee replied:
앗 그렇군요 더 학습해보겠습니다 🫡

yebink commented on src/main/java/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,80 @@
+public class AttendanceRecord {
+    private static final int DATE_INDEX = 0;
+    private static final int TIME_INDEX = 1;
+    public static final LocalTime ABSENT_TIME = LocalTime.of(14, 0);
```
결석에 대한 데이터를 저장하기 위해 임의의 시간을 설정하셨네요!
null 이 아닌 임의의 시간을 지정한 이유는 무엇이고, 많고 많은 시간 중에 14:00로 설정하신 이유는 무엇인가요?

yesjuhee replied:
`null`값은 최대한 지양하는게 좋다고 알고 있어서 임의의 시간을 지정했습니다!
월요일은 등교 시간이 1시이기 때문에 모든 요일에 결석으로 기록될 수 있는 시간을 지정하기 위해 14시를 선택했습니다.
14시를 하드코딩하면 의미 전달이 안될 것 같아서 상수로 정의했습니다.

yebink replied:
그러한 이유였군요! 😄
null 을 지양하고자 하는 의견에 동의하지만,
임의의 시간을 지정할 경우, 누군가 오해하는 일이 발생할 수도 있지 않을까요?

필요하면 사용하되, NPE가 발생하지 않도록 잘 방어하는 연습을 해보는 것도 중요합니다 👍

yesjuhee replied:
오해의 소지가 충분하네요!
다른 피드백에서 남겨주신 것처럼 기록이 없는 경우 따로 데이터를 저장하지 않도록 리팩터링하면 이 문제를 해결할 수 있을 것 같습니다.
좋은 의견 감사합니다 👍 👍

yebink commented on src/main/java/domain/CurrentDateGenerator.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.time.LocalDate;
+
+public class CurrentDateGenerator implements DateGenerator {
```
미사용 클래스네요!

yesjuhee replied:
구현 방향이 바뀌면서 삭제해야 했는데 깜빡했네요
감사합니다 😄

yesjuhee replied:
80d03a0

yebink commented on src/main/java/domain/CrewAttendanceRecords.java:
```diff
@@ -0,0 +1,116 @@
+        crews.sort(Comparator.comparing(crew -> {
+            int absentCount = crewAttendanceRecords.get(crew).getAbsentCount();
+            int tardyCount = crewAttendanceRecords.get(crew).getTardyCount();
+            absentCount += (tardyCount / 3);
+            return (absentCount + tardyCount % 3) * -1;
```
`모든 원시 값과 문자열을 포장한다` 요구사항을 지켜볼까요?

yesjuhee replied:
da45820

yebink commented:
안녕하세요 노랑~ 
출석 미션 리뷰를 맡은 엘리입니다 😀 

주신 질문에 대한 답변 드려요.

**1. 모든 원시값을 포장하기**
* `String` 을 `Crew` 로 포장하신 것은 충분히 의미 있는 선택이라고 봅니다 💯 
* 지금 구조에서는 큰 효용을 못 느끼실 수 있으나, 신규 요구사항에 유연하게 대응할 수 있겠네요.
  * (그럴 일은 없겠지만) `Crew`의 나이에 따라 지각 조건을 다르게 설정하는 요구사항이 추가되면 어떨까요?
* 그렇다고 모든 원시값을 무조건 객체로 포장할 필요는 없으니 노랑만의 판단 기준을 세워보세요 🙂 

**2. 시작 날짜 입력 (`DateGenerator`)**
* `DateGenerator` 인터페이스는 잘 정의하셨습니다! 그런데 구현체가 모두 람다식이네요 ㅎㅎ
* 람다식은 재사용이 어렵고, 상태를 가지지 않는 등등의 이유로 객체지향과는 거리가 먼 친구입니다.
  * 개인적으로도 람다식은 코드를 이해하기 어려워서 좋아하지 않고, 잘 사용하지 않습니다 😢 
  * 노랑이 람다식을 사용하면서 느꼈던 장점은 무엇이었나요~?
* 람다가 아닌 구현 클래스로도 작성해보시고, 둘의 차이점을 경험해보시면 좋을 것 같습니다!

**3. 팩터리 메서드 패턴 (`CsvParsingGenerator`)**
* 네 팩터리 메서드 패턴대로 잘 구현하셨습니다!
* 한가지 개선사항이 있다면, `CsvParsingGenerator` 말고 다른 구현체도 만들어볼까요?
* 현재 테스트에서도 `CsvParsingGenerator` 를 사용하기 때문에, 테스트마다 모두 동일한 csv 파일을 읽고 있습니다. 
* 테스트마다 필요한 정보만 읽을 수 있도록 다른 구현체도 생성해보아요 🙂 

**4. 정적 팩터리 메서드**
* 네 잘 작성하셨습니다~ 💯 
* 다만 `parse` 와 `checkIn` 메서드를 AttendanceRecord 에서 꼭 가져야 할지 고민해보세요~

yebink reviewed (changes requested):
안녕하세요 노랑~

출석 미션 구현하시느라 고생하셨습니다 👍 
코멘트 확인 부탁드려요~!

yesjuhee committed:
refactor : 메서드명 변경

yesjuhee committed:
refactor : equals 메서드를 사용하여 enum을 비교하는 경우 `==`로 수정

yesjuhee committed:
refactor : modifier 순서 변경, 불필요한 괄호 삭제, import 정리

yesjuhee committed:
refactor : 매직넘버 상수화

yesjuhee committed:
refactor : DateGenerator 인터페이스 삭제

- 인터페이스를 사용하지 않고 LocalDate 값을 바로 넣어주도록 리팩터링

yesjuhee commented:
>  2. 시작 날짜 입력 (DateGenerator)
> 
> DateGenerator 인터페이스는 잘 정의하셨습니다! 그런데 구현체가 모두 람다식이네요 ㅎㅎ
> 람다식은 재사용이 어렵고, 상태를 가지지 않는 등등의 이유로 객체지향과는 거리가 먼 친구입니다.
> 개인적으로도 람다식은 코드를 이해하기 어려워서 좋아하지 않고, 잘 사용하지 않습니다 😢
> 노랑이 람다식을 사용하면서 느꼈던 장점은 무엇이었나요~?
> 람다가 아닌 구현 클래스로도 작성해보시고, 둘의 차이점을 경험해보시면 좋을 것 같습니다!

리팩터링과정에서 다른 크루와 논의하면서 인터페이스를 남발하고 있지 않은가에 대해 고민해봤습니다.
곰곰이 다시 생각해보면서 `DateGenerator`라는 인터페이스가 없이 그냥 기준이 되는 `LocalDate`값을 주입해주면 된다고 판단을 해서 `DateGenerator`를 사용하지 않는 방향으로 리팩터링했습니다.

80d03a0

yesjuhee committed:
refactor(attendance_record) : 중복된 메서드 삭제

- AttendanceRecord의 checkIn 메서드 삭제

yesjuhee committed:
refactor(attendance_record) : 정적팩터리메서드 parse 삭제

- AttendanceRecord의 parse 메서드 삭제
- CsvParsingGenerator에 출석 기록 파싱 책임 역할 부여

yesjuhee committed:
refactor(test) : 테스트 코드 리팩터링

- AttendanceRecord의 parse 메서드 삭제 반영
- 중복으로 사용되는 로컬 변수를 멤버 변수로 분리
- given-when-then 구조로 변경

yesjuhee committed:
chore : 사용되지 않는 변수, 중복 사용된 메서드 정리

yesjuhee committed:
refactor(crew_attendance_record_generator) : 테스트를 위한 구현체 추가

- CrewAttendanceRecordsGenerator 인터페이스의 구현체 다양화
- CsvParsingGenerator 프로덕션 코드에서만 사용하도록 리팩터링
- BingTeeAttendanceRecordsGenerator - 빙티의 출석 기록을 만들어주는 테스트용 구현체
- WarnedCrewRecordsGenerator - 제적 위험자 반환 기능을 테스트 하기 위한 구현체

yesjuhee commented:
> 팩터리 메서드 패턴 (CsvParsingGenerator)
> 
> 네 팩터리 메서드 패턴대로 잘 구현하셨습니다!
> 한가지 개선사항이 있다면, CsvParsingGenerator 말고 다른 구현체도 만들어볼까요?
> 현재 테스트에서도 CsvParsingGenerator 를 사용하기 때문에, 테스트마다 모두 동일한 csv 파일을 읽고 있습니다.
> 테스트마다 필요한 정보만 읽을 수 있도록 다른 구현체도 생성해보아요 🙂

0644962

리뷰해주신대로 테스트를 위한 구현체를 따로 구현하여 리팩터링을 했습니다!
확실히 구현체를 분리해두니 테스트를 위한 객체 조작이 쉬워지고 테스트가 편해지는 것을 느낄 수 있었습니다.

---
`CrewAttendnaceRecordsTest`의 `getWarnedCrewsTest()`의 경우 아래 요구사항을 테스트하기 위해 다양한 상태를 가지는 테스트용 객체가 필요하기 때문에 이 테스트를 위한 `WarnedCrewRecordsGenerator`를 추가로 작성했습니다.
> 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 출석 상태가 같으면 닉네임으로 오름차순 정렬한다.

그런데 해당 generator를 보시면 알겠지만 원하는 상태의 데이터를 만들기 위해서 반복되는 코드도 많고 변수명이나 메서드 명에 숫자가 들어가는 등 좀 지저분하게 작성이 됐는데요. 만약 이 부분을 개선하려면 **어떻게 코드를 수정하는게 좋을까요?**

또한 테스트를 위한 구현체는 `test/java/domain` 패키지에 위치시켰는데 이 위치가 적절한지, 또한 generator를 위한 테스트 코드를 추가로 작성하는게 좋을지도 여쭙고 싶습니다!

yebink commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public Crew readNickname() {
+        System.out.println("\n닉네임을 입력해 주세요.");
```
`System.lineSeparator()` 를 사용하면 OS 에 구애받지 않고 줄바꿈이 가능합니다.

```suggestion
        System.out.println(System.lineSeparator() + "닉네임을 입력해 주세요.");
```

yebink reviewed (approved):
안녕하세요 노랑~

피드백 잘 반영해주셨네요 💯 
추가 피드백은 step2에서 반영해주세요!
머지하겠습니다 ~

yebink commented:
- **WarnedCrewRecordsGenerator 테스트 관련**
  - 잘 리팩토링 하셨습니다! 조금 더 가독성을 높여본다면, 테스트 픽스쳐를 활용할 수 있을 것 같아요.
  - 특정 날짜 / 시간 / 분 에 따라 AttendanceRecord 를 생성하는 fixture를 만들고, 
  - AttendanceRecords 에 가변인자 생성자를 추가해 add 없이 바로 생성합니다.
  - 그리고 보통 이런 데이터 셋팅을 위한 클래스는 `test/java/fixture` 하위에 둡니다!
  - 방법은 다양하니 아래 코드는 참고만 해주세요~
```java
public class WarnedCrewRecordsGenerator implements CrewAttendanceRecordsGenerator {
    // 생략
    private void fillWarningCrew() {
        Crew crew = new Crew("경고크루");
        AttendanceRecords attendanceRecords = new AttendanceRecords(
                AttendanceRecordFixture.getAbsent(2),
                AttendanceRecordFixture.getAbsent(3)
        );
        warnedCrewRecords.put(crew, attendanceRecords);
    }

    // 생략
    static class AttendanceRecordFixture {

        static AttendanceRecord getAbsent(int day) {
            return AttendanceRecord.asAbsent(LocalDate.of(2024, 12, day));
        }

        static AttendanceRecord of(int day, int hour, int minute) {
            return AttendanceRecord.of(LocalDate.of(2024, 12, day), LocalTime.of(hour, minute));
        }
    }
}
```
